File: hazama/ui/editor.py
from PySide.QtGui import *
from PySide.QtCore import *
from hazama.ui.editor_ui import Ui_editor
from hazama.ui.customobjects import TagCompleter
from hazama.ui.customwidgets import DateTimeDialog
from hazama.ui import (font, datetimeTrans, currentDatetime, fullDatetimeFmt,
                       saveWidgetGeo, restoreWidgetGeo, datetimeToQt, dbDatetimeFmtQt)
from hazama.config import settings, db


class Editor(QWidget, Ui_editor):
    """The widget that used to edit diary's body, title, tag and datetime.
    Signal closed: (id of nikki, needSave)
    """
    closed = Signal(int, bool)

    def __init__(self, nikkiDict, parent=None):
        super().__init__(parent)
        self._saveOnClose = True
        self.setupUi(self)
        self.readOnly = self.datetime = self.id = self.timeModified = self.tagModified = None
        restoreWidgetGeo(self, settings['Editor'].get('windowGeo'))

        self.titleEditor.setFont(font.title)
        self.titleEditor.returnPressed.connect(
            lambda: None if self.readOnly else self.textEditor.setFocus())
        self.textEditor.setFont(font.text)
        self.textEditor.setAutoIndent(settings['Editor'].getboolean('autoIndent'))

        self.dtBtn.setFont(font.datetime)
        sz = max(font.datetime_m.ascent(), 12)
        self.dtBtn.setIconSize(QSize(sz, sz))
        self.lockBtn.setIconSize(QSize(sz, sz))
        self.lockBtn.clicked.connect(lambda: self.setReadOnly(False))

        self.tagEditor.setTextMargins(QMargins(2, 0, 2, 0))
        self.tagEditor.setCompleter(TagCompleter(list(db.gettags()), self.tagEditor))
        self.tagEditor.returnPressed.connect(
            lambda: None if self.readOnly else self.box.button(QDialogButtonBox.Save).setFocus())

        # setup shortcuts
        # seems PySide has problem with QKeySequence.StandardKeys
        self.closeSaveSc = QShortcut(QKeySequence.Save, self, self.close)
        self.closeNoSaveSc = QShortcut(QKeySequence('Ctrl+W'), self, self.closeNoSave)
        self.quickCloseSc = QShortcut(QKeySequence('Esc'), self, self.closeNoSave)
        # Ctrl+Shift+Backtab doesn't work
        self.preSc = QShortcut(QKeySequence('Ctrl+Shift+Tab'), self)
        self.quickPreSc = QShortcut(QKeySequence('Left'), self)
        self.nextSc = QShortcut(QKeySequence('Ctrl+Tab'), self)
        self.quickNextSc = QShortcut(QKeySequence('Right'), self)

        self.fromNikkiDict(nikkiDict)

    def showEvent(self, event):
        if settings['Editor'].getboolean('titleFocus'):
            self.titleEditor.setCursorPosition(0)
        else:
            self.textEditor.setFocus()
            self.textEditor.moveCursor(QTextCursor.Start)
        event.accept()

    def closeEvent(self, event):
        """Normal close will save diary. For cancel operation, call closeNoSave."""
        settings['Editor']['windowGeo'] = saveWidgetGeo(self)
        self.closed.emit(self.id, self.needSave() if self._saveOnClose else False)
        event.accept()

    def mousePressEvent(self, event):
        """Handle mouse back/forward button"""
        if event.button() == Qt.XButton1:  # back
            self.preSc.activated.emit()
            event.accept()
        elif event.button() == Qt.XButton2:  # forward
            self.nextSc.activated.emit()
            event.accept()
        else:
            super().mousePressEvent(event)

    def closeNoSave(self):
        self._saveOnClose = False
        self.close()

    def needSave(self):
        return (self.textEditor.document().isModified() or
                self.titleEditor.isModified() or self.timeModified or
                self.tagModified)

    def setReadOnly(self, readOnly):
        for i in [self.titleEditor, self.textEditor, self.tagEditor]:
            i.setReadOnly(readOnly)
        self.dtBtn.setCursor(Qt.ArrowCursor if readOnly else Qt.PointingHandCursor)
        self.box.setStandardButtons(self.box.Close if readOnly else
                                    self.box.Save | self.box.Cancel)
        self.box.button(self.box.Close if readOnly else self.box.Save).setDefault(True)
        self.lockBtn.setVisible(readOnly)
        self.titleEditor.setVisible(not readOnly or bool(self.titleEditor.text()))
        self.tagEditor.setVisible(not readOnly or bool(self.tagEditor.text()))
        for i in [self.quickCloseSc, self.quickPreSc, self.quickNextSc]:
            i.setEnabled(readOnly)
        self.readOnly = readOnly

    def fromNikkiDict(self, dic):
        self.timeModified = self.tagModified = False
        self.id = dic['id']
        self.datetime = dic.get('datetime')

        self.dtBtn.setText(datetimeTrans(self.datetime) if self.datetime else '')
        self.titleEditor.setText(dic.get('title', ''))
        self.tagEditor.setText(dic.get('tags', ''))
        self.textEditor.setRichText(dic.get('text', ''), dic.get('formats'))
        # if title is empty, use datetime instead. if no datetime (new), use "New Diary"
        t = (dic.get('title') or
             (datetimeTrans(self.datetime, stripTime=True) if 'datetime' in dic else None) or
             self.tr('New Diary'))
        self.setWindowTitle("%s - Hazama" % t)

        readOnly = (settings['Editor'].getboolean('autoReadOnly') and
                    self.datetime is not None and
                    datetimeToQt(self.datetime).daysTo(QDateTime.currentDateTime()) > 3)
        self.setReadOnly(readOnly)

    def toNikkiDict(self):
        text, formats = self.textEditor.getRichText()
        return dict(id=self.id, datetime=self.datetime or currentDatetime(),
                    text=text, formats=formats, title=self.titleEditor.text(),
                    tags=self.tagEditor.text())

    @Slot()
    def on_tagEditor_textEdited(self):
        # tagEditor.isModified() will be reset by completer. So this instead.
        self.tagModified = True

    @Slot()
    def on_dtBtn_clicked(self):
        """Show datetime edit dialog"""
        if self.readOnly: return
        dtStr = currentDatetime() if self.datetime is None else self.datetime
        newDt = DateTimeDialog.getDateTime(datetimeToQt(dtStr), fullDatetimeFmt, self)
        if newDt is not None:
            newDtStr = newDt.toString(dbDatetimeFmtQt)
            if newDtStr != self.datetime:
                self.datetime = newDtStr
                self.dtBtn.setText(datetimeTrans(newDtStr))
                self.timeModified = True

File: hazama/editor.cpp
#include "editor.h"
#include "QShortcut"
#include "globals.h"
#include "customobjects/nshadoweffect.h"
#include "customobjects/datetimedialog.h"

#ifdef Q_OS_WIN
#include <windows.h>
#include <windowsx.h>
#include <QtWinExtras/QtWin>
#endif


Editor::Editor(QWidget *parent) : QFrame(parent) {
    setupUi(this);
    lockBtn->hide();
    restoreWidgetGeo(this, settings->value("Editor/windowGeo").toString());

    titleEditor->setFont(fonts.title);
    connect(titleEditor, &QLineEdit::returnPressed, [&](){
        if (!readOnly)
            textEditor->setFocus();
    });
    textEditor->setFont(fonts.text);
    textEditor->setAutoIndent(settings->value("Editor/autoIndent").toBool());
    textEditor->setTabChangesFocus(!settings->value("Editor/tabIndent").toBool());

    tagEditor->setTextMargins(QMargins(2, 0, 2, 0));
    //tagEditor->setCompleter(TagCompleter(list(db.get_tags()), self.tagEditor));
    connect(tagEditor, &QLineEdit::returnPressed, [&](){
        if (!readOnly)
            box->button(QDialogButtonBox::Save)->setFocus();
    });
    // tagEditor.isModified() will be reset by completer. Use this instead.
    connect(tagEditor, &QLineEdit::textEdited, [&](){ tagModified = true; });

    dtBtn->setFont(fonts.datetime);
    int sz = qMax(QFontMetrics(fonts.datetime).ascent(), 12);
    dtBtn->setIconSize(QSize(sz, sz));
    lockBtn->setIconSize(QSize(sz, sz));
    connect(lockBtn, &QPushButton::clicked, [&](){ setReadOnly(false); });

    if (osInfo.isWin10 && settings->value("extendTitleBarBg").toBool())
        bottomArea->setProperty("bgType", "win10");

    //// setup shortcuts
    // seems PySide has problem with QKeySequence.StandardKeys
    closeSaveSc = new QShortcut(QKeySequence::Save, this);
    connect(closeSaveSc, &QShortcut::activated, this, &Editor::close);
    closeNoSaveSc = new QShortcut(QKeySequence("Ctrl+W"), this);
    connect(closeNoSaveSc, &QShortcut::activated, this, &Editor::closeNoSave);
    quickCloseSc = new QShortcut(QKeySequence("Esc"), this);
    connect(quickCloseSc, &QShortcut::activated, this, &Editor::closeNoSave);
    // Ctrl+Shift+Backtab doesn't work
    preSc = new QShortcut(QKeySequence("Ctrl+Shift+Tab"), this);
    quickPreSc = new QShortcut(QKeySequence("Left"), this);
    nextSc = new QShortcut(QKeySequence("Ctrl+Tab"), this);
    quickNextSc = new QShortcut(QKeySequence("Right"), this);
}

void Editor::showEvent(QShowEvent *e) {
    Q_UNUSED(e)
    applyExtendTitleBarBg();
    if (settings->value("Editor/titleFocus").toBool())
        titleEditor->setCursorPosition(0);
    else {
        textEditor->setFocus();
        textEditor->moveCursor(QTextCursor::Start);
    }
}

void Editor::closeEvent(QCloseEvent *e) {
    Q_UNUSED(e)
    settings->setValue("Editor/windowGeo", saveWidgetGeo(this));
    emit closed(id, (noSave) ? false : isModified());
}

#ifdef Q_OS_WIN
bool Editor::nativeEvent(const QByteArray &eventType, void *message, long *result) {
    if (eventType != QByteArrayLiteral("windows_generic_MSG"))
        return false;
    auto msg = static_cast<MSG*>(message);
    if (msg->message == WM_NCHITTEST) {
        auto globalPos = QPoint(GET_X_LPARAM(msg->lParam), GET_Y_LPARAM(msg->lParam));
        auto pos = mapFromGlobal(globalPos);
        auto widget = childAt(pos);
        if (widget == bottomArea) {
            *result = HTCAPTION;
            return true;
        }
    }
    return false;
}
#endif

// Handle mouse back/forward button
void Editor::mousePressEvent(QMouseEvent *event) {
    if (event->button() == Qt::XButton1) {  // back
        emit preSc->activated();
        event->accept();
    } else if (event->button() == Qt::XButton2) {  // forward
        emit nextSc->activated();
        event->accept();
    } else
        QFrame::mousePressEvent(event);
}

void Editor::applyExtendTitleBarBg() {
    if (settings->value("extendTitleBarBg").toBool()) {
#ifdef Q_OS_WIN
        if (!QtWin::isCompositionEnabled())
            return;
        QtWin::extendFrameIntoClientArea(this, 0, 0, 0, bottomArea->height());
        setAttribute(Qt::WA_TranslucentBackground);
        // Qt5 only hack, Qt4 doesn't need this
        setAttribute(Qt::WA_NoSystemBackground, false);  // it's set by WA_TranslucentBackground
        // force background of MainWindow to be cleared before its contents being painted
        QPalette pal = palette();
        pal.setColor(QPalette::Background, Qt::transparent);
        setPalette(pal);
#endif
        for (auto i : {dtBtn, lockBtn}) {
            auto eff = new NShadowEffect((osInfo.isWin7) ? 5 : 3, i);
            eff->setColor(QColor(Qt::white));
            eff->setOffset(0, 0);
            eff->setBlurRadius(((osInfo.isWin7) ? 16 : 8) * scaleRatio);
            i->setGraphicsEffect(eff);
        }
    } else {
#ifdef Q_OS_WIN
        QtWin::resetExtendedFrame(this);
        setPalette(QApplication::palette(this));  // set palette to default one
#endif
        for (auto i : {dtBtn, lockBtn}) {
            auto eff = i->graphicsEffect();
            if (eff) {
                i->setGraphicsEffect(nullptr);
                delete eff;
            }
        }
    }
}

void Editor::fromDiary(const Diary &d) {
    timeModified = tagModified = false;
    id = d.id;
    datetime = d.datetime;
    dtBtn->setText((datetime.isNull()) ? "" : datetimeTrans(datetime));
    titleEditor->setText(d.title);
    tagEditor->setText(d.tags.join(" "));
    textEditor->setRichText(d.text, d.formats);
    // if title is empty, use datetime instead. if no datetime (new), use "New Diary"
    auto t = d.title;
    if (t.isEmpty() && !datetime.isNull())
        t = dtBtn->text();
    if (t.isEmpty())
        t = tr("New Diary");
    setWindowTitle(QString("%1 - Hazama").arg(t));

    setReadOnly(settings->value("Editor/autoReadOnly").toBool() &&
                !datetime.isNull() &&
                datetime.daysTo(QDateTime::currentDateTime()) > 3);
}

Diary Editor::toDiary() {
    auto rt = textEditor->getRichText();
    auto tags = tagEditor->text().split(" ", QString::SkipEmptyParts);
    if (tagModified)
        tags.removeDuplicates();
    auto dt = (datetime.isNull()) ? QDateTime::currentDateTime() : datetime;
    return { id, dt, rt.first, titleEditor->text(), tags, rt.second };
}

bool Editor::isModified() {
    return textEditor->document()->isModified() ||
           titleEditor->isModified() || timeModified || tagModified;
}

void Editor::setReadOnly(bool to) {
    readOnly = to;
    titleEditor->setReadOnly(to);
    textEditor->setReadOnly(to);
    tagEditor->setReadOnly(to);
    dtBtn->setCursor((to) ? Qt::ArrowCursor : Qt::PointingHandCursor);
    box->setStandardButtons((to) ? QDialogButtonBox::Close :
                            QDialogButtonBox::Save | QDialogButtonBox::Cancel);
    box->button((to) ? QDialogButtonBox::Close : QDialogButtonBox::Save)->setDefault(true);
    lockBtn->setVisible(to);
    titleEditor->setVisible(!to || !titleEditor->text().isEmpty());
    tagEditor->setVisible(!to || !tagEditor->text().isEmpty());
    for (auto i : {quickCloseSc, quickPreSc, quickNextSc})
        i->setEnabled(to);

#ifdef Q_OS_WIN
    if (settings->value("extendTitleBarBg").toBool())  // height may change
        QtWin::extendFrameIntoClientArea(this, 0, 0, 0, bottomArea->height());
#endif
}

void Editor::editDateTime() {
    if (readOnly)
        return;
    auto dt = (datetime.isNull()) ? QDateTime::currentDateTime() : datetime;
    auto newDt = DateTimeDialog::getDateTime(dt, fullDatetimeFmt, this);
    if (!newDt.isNull() && newDt != dt) {
        datetime = newDt;
        dtBtn->setText(datetimeTrans(newDt));
        timeModified = true;
    }
}
